Match protobuf to py-protobuf version

bvanessen opened the pull request:
Fixed the py-protobuf recipe so that when cpp support is require so
that it uses the same major and minor version range for the protobuf
package.

bvanessen committed:
Fixed the py-protobuf recipe so that when cpp support is require so
that it uses the same major and minor version range for the protobuf
package.

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -71,7 +71,16 @@ class PyProtobuf(PythonPackage):
+    # Setup dependencies for protobuf to use the same minor version as py-protobuf
+    # Handle the 3.x series releases
+    for ver in list(range(1, 20)):
+        depends_on("protobuf@3." + str(ver) + ".0:3." + str(ver) + ".99",
+                   type=("build", "run"),
+                   when="+cpp @3." + str(ver) + ".0:3." + str(ver) + ".99")
```
```suggestion
        depends_on("protobuf@3." + str(ver),
                   type=("build", "run"),
                   when="+cpp @3." + str(ver))
```

bvanessen replied:
@adamjstewart So there isn't always a one for one patch version that corresponds between the two packages.

adamjstewart replied:
Yes, `@1.2` is equivalent to `@1.2.0:1.2.99` but technically more correct because it works even if there is a `1.2.100` release.

bvanessen replied:
@adamjstewart Okay thanks.

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -71,7 +71,16 @@ class PyProtobuf(PythonPackage):
+                   type=("build", "run"),
+                   when="+cpp @3." + str(ver) + ".0:3." + str(ver) + ".99")
+    # Handle the 2.x series releases
+    depends_on("protobuf@2.0.0:2.99.99", type=("build", "run"),
+               when="+cpp @2.0.0:2.99.99")
```
```suggestion
    depends_on("protobuf@2", type=("build", "run"),
               when="+cpp @2")
```

bvanessen committed:
Fixed the range bound for the 3.x py-protobuf packages.

Added mappings for 4.x py-protobuf packages to 3.x protobuf packages.

Removed a hash for v21.1 protobuf and replaced with v3.21.1 to keep a
standard versioning convention.  Note that while Google has started
releasing both 3.x.y and a tag that dropped the leading 3. so it is
just x.y.  This provides the appearance of a new major version, but
really is just a new minor version.  These packages still report
versions as 3.x.y, so switching to versions and hashes with that
convention.

bvanessen committed:
Simplified constraints based on reviewer comments.

bvanessen committed:
Fixed flake8 errors

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -19,6 +19,7 @@ class PyProtobuf(PythonPackage):
 
     variant("cpp", default=False, description="Enable the cpp implementation")
 
+    version("4.21.5", sha256="eb1106e87e095628e96884a877a51cdb90087106ee693925ec0a300468a9be3a")
```
Note that this new version is likely to break many packages with open-ended version dependencies. See https://github.com/spack/spack/pull/30880 for details. This isn't a reason not to add it, just a note that we should expect to see build issue reports due to this.

bvanessen replied:
Why don't I either remove it or put a preferred on it to avoid it.

adamjstewart replied:
Up to you. The packages I care about have already been fixed. Most developers have discovered this issue already and locked the maximum version in their dependency list, so packages with a new release since then that have been added to Spack shouldn't be affected. Personally, I would rather get the bug reports so we can fix this. But I realize that's not the most stable solution. Another option would be to lock all packages to `protobuf@:3` until someone tests and verifies that they work with protobuf 4.

bvanessen commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -19,6 +19,7 @@ class PyProtobuf(PythonPackage):
     variant("cpp", default=False, description="Enable the cpp implementation")
 
+    version("4.21.5", sha256="eb1106e87e095628e96884a877a51cdb90087106ee693925ec0a300468a9be3a")
     version("3.20.1", sha256="adc31566d027f45efe3f44eeb5b1f329da43891634d61c75a5944e9be6dd42c9")
```
```suggestion
    version("3.20.1", sha256="adc31566d027f45efe3f44eeb5b1f329da43891634d61c75a5944e9be6dd42c9", preferred=True)
```

bvanessen replied:
@adamjstewart Do you want to merge my suggestion to see if it goes through without forcing a new CI run and approval process?

adamjstewart replied:
Let's see if the current solution passes CI. If it does, we can merge without pinning

bvanessen committed:
Update var/spack/repos/builtin/packages/py-protobuf/package.py

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -71,7 +72,16 @@ class PyProtobuf(PythonPackage):
+    for ver in list(range(1, 21)):
+        depends_on("protobuf@3." + str(ver), type=("build", "run"), when="+cpp @3." + str(ver))
+    # Handle the 2.x series releases
+    depends_on("protobuf@2", type=("build", "run"), when="+cpp @2")
```
Out of curiosity, where did you find these mappings? 3 and 4 are somewhat intuitive, but is there official docs somewhere that state that any 2.x version is compatible with any 2.x version?

bvanessen replied:
@adamjstewart No I haven't found official documentation.  However, for the 2.x packages, there isn't a matching set of versions in both recipes.

bvanessen replied:
@adamjstewart There is a matching version so I tightened up the constraint.

bvanessen committed:
Fixed constraints on v2. versions and addressed Flake8 comments.

bvanessen committed:
Merge branch 'bugfix_py_protobuf_deps' of github.com:bvanessen/spack into bugfix_py_protobuf_deps

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -71,7 +75,16 @@ class PyProtobuf(PythonPackage):
+    for ver in list(range(1, 21)):
+        depends_on("protobuf@3." + str(ver), type=("build", "run"), when="+cpp @3." + str(ver))
+    # Handle the 2.x series releases
+    depends_on("protobuf@2.", type=("build", "run"), when="+cpp @2.")
```
Did you mean to add a for-loop here?

bvanessen replied:
@adamjstewart Yes, thanks.

bvanessen committed:
Fixed flake8

bvanessen committed:
Fixed range dependencies for version 2.x

adamjstewart commented on var/spack/repos/builtin/packages/py-protobuf/package.py:
```diff
@@ -19,7 +19,12 @@ class PyProtobuf(PythonPackage):
+    version(
+        "3.20.1",
+        sha256="adc31566d027f45efe3f44eeb5b1f329da43891634d61c75a5944e9be6dd42c9",
+        preferred=True
```
```suggestion
        preferred=True,
```

bvanessen committed:
Update var/spack/repos/builtin/packages/py-protobuf/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

bvanessen commented:
@adamjstewart Do you know what is going on with the radius CI check that failed?  I am not familiar with it.

adamjstewart commented:
Seems transient, I think the next time it runs it will work.

You should worry more about the package audits:
```
1. py-protobuf: dependency on protobuf@3.8 cannot be satisfied by known versions of protobuf
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
    known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.17.3, 3.17.0, 3.[16](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:17).0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
2. py-protobuf: dependency on protobuf@2.3 cannot be satisfied by known versions of protobuf
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
    known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.[17](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:18).3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
3. py-protobuf: dependency on protobuf@2.4 cannot be satisfied by known versions of protobuf
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
    known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.17.3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
4. py-protobuf: dependency on protobuf@2.6 cannot be satisfied by known versions of protobuf
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
    known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.[18](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:19).2, 3.18.0, 3.17.3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
```

bvanessen committed:
Fixed version ranges to skip unknown versions.

bvanessen committed:
Merge branch 'bugfix_py_protobuf_deps' of github.com:bvanessen/spack into bugfix_py_protobuf_deps

bvanessen commented:
> Seems transient, I think the next time it runs it will work.
> 
> You should worry more about the package audits:
> 
> ```
> 1. py-protobuf: dependency on protobuf@3.8 cannot be satisfied by known versions of protobuf
>     happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
>     known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.17.3, 3.17.0, 3.[16](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:17).0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
> 2. py-protobuf: dependency on protobuf@2.3 cannot be satisfied by known versions of protobuf
>     happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
>     known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.[17](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:18).3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
> 3. py-protobuf: dependency on protobuf@2.4 cannot be satisfied by known versions of protobuf
>     happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
>     known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.18.2, 3.18.0, 3.17.3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
> 4. py-protobuf: dependency on protobuf@2.6 cannot be satisfied by known versions of protobuf
>     happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-protobuf/package.py
>     known versions of protobuf are 3.21.5, 3.21.4, 3.21.3, 3.21.2, 3.21.1, 3.20.1, 3.20.0, 3.19.4, 3.19.3, 3.19.2, 3.[18](https://github.com/spack/spack/runs/8235497837?check_suite_focus=true#step:5:19).2, 3.18.0, 3.17.3, 3.17.0, 3.16.0, 3.15.7, 3.15.5, 3.15.4, 3.15.2, 3.15.1, 3.15.0, 3.14.0, 3.13.0, 3.12.3, 3.12.2, 3.12.1, 3.12.0, 3.11.4, 3.11.3, 3.11.2, 3.11.1, 3.11.0, 3.10.1, 3.10.0, 3.9.2, 3.9.1, 3.7.1, 3.7.0, 3.6.1, 3.5.2, 3.5.1.1, 3.5.1, 3.5.0.1, 3.5.0, 3.4.1, 3.4.0, 3.3.0, 3.2.0, 3.1.0, 3.0.2, 2.5.0
> ```

@adamjstewart Done.

bvanessen committed:
Fixed the dependencies on protobuf to solve weird build issues.
